chain/src/chain_service.rs

```diff
@@ -0,0 +1,247 @@
+        let signal_receiver = new_crossbeam_exit_rx();
+
+        loop {
+            select! {
+                recv(self.process_block_rx) -> msg => match msg {
+                    Ok(Request { responder, arguments: lonely_block }) => {
+                        // asynchronous_process_block doesn't interact with tx-pool,
+                        // no need to pause tx-pool's chunk_process here.
+                        self.asynchronous_process_block(lonely_block);
+                        let _ = responder.send(());
+                    },
+                    _ => {
+                        error!("process_block_receiver closed");
+                        break;
+                    },
+                },
+                recv(signal_receiver) -> _ => {
+                    info!("ChainService received exit signal, exit now");
+                    break;
+                }
+            }
+        }
+    }
+
+    fn non_contextual_verify(&self, block: &BlockView) -> Result<(), Error> {
+        let consensus = self.shared.consensus();
+        BlockVerifier::new(consensus).verify(block).map_err(|e| {
+            debug!("[process_block] BlockVerifier error {:?}", e);
+            e
+        })?;
+
+        NonContextualBlockTxsVerifier::new(consensus)
+            .verify(block)
+            .map_err(|e| {
+                debug!(
+                    "[process_block] NonContextualBlockTxsVerifier error {:?}",
+                    e
+                );
+                e
+            })
+            .map(|_| ())
+    }
+
+    // make block IO and verify asynchronize
+    fn asynchronous_process_block(&self, lonely_block: LonelyBlockWithCallback) {
+        let block_number = lonely_block.block().number();
+        let block_hash = lonely_block.block().hash();
+        if block_number < 1 {
+            warn!("receive 0 number block: 0-{}", block_hash);
+        }
+
+        if lonely_block.switch().is_none()
+            || matches!(lonely_block.switch(), Some(switch) if !switch.disable_non_contextual())
+        {
+            let result = self.non_contextual_verify(lonely_block.block());
+            if let Err(err) = result {
+                error!(
+                    "block {}-{} verify failed: {:?}",
+                    block_number, block_hash, err
+                );
+                self.shared
+                    .insert_block_status(lonely_block.block().hash(), BlockStatus::BLOCK_INVALID);
+                tell_synchronizer_to_punish_the_bad_peer(
+                    self.verify_failed_blocks_tx.clone(),
+                    lonely_block.peer_id_with_msg_bytes(),
+                    lonely_block.block().hash(),
+                    &err,
+                );
+
+                lonely_block.execute_callback(Err(err));
+                return;
+            }
+        }
+
+        match self.lonely_block_tx.send(lonely_block) {
+            Ok(_) => {}
+            Err(SendError(lonely_block)) => {
```

Comment:
What we can do here is to determine whether the channel is closed. If it is closed, we can only panic. If it is not closed, we can try to resend it.

Reply:
When the ConsumeOrphan thread receives a stop signal (Ctrl-C), the receiver `lonely_block_rx` will be dropped. Consequently, if there is an attempt to send `lonely_block` using `self.lonely_block_tx.send(lonely_block)`, it may result in the `Err(SendError(..))` error. This situation is indeed possible.

And I am concerned that the panic message may alarm or confuse users.